File: dp/22.generate-parentheses.cpp
/*
 * @lc app=leetcode id=22 lang=cpp
 *
 * [22] Generate Parentheses
 */
#include "bits/stdc++.h"

using namespace std;

// @lc code=start
class Solution {
  vector<string> res;
  string comb;

public:
  vector<string> generateParenthesis(int n) {

    backtrack(n, n);
    return res;
  }
  void backtrack(int left, int right) {
    if (left) {
      comb += "(";
      backtrack(left - 1, right);
      comb.pop_back();
    }

    if (right > left) {
      comb += ")";
      backtrack(left, right - 1);
      comb.pop_back();
    }
    if (!right)
      res.push_back(comb);
  }
};
// @lc code=end

int main() {
  Solution s;
  for (const auto val : s.generateParenthesis(3)) {
    cout << val << '\n';
  }

  cout << endl;
}

File: dp/22.generate-parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#

# @lc code=start
from functools import cache
from typing import List

# TODO: rescond review
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
      def backtrack(p, left, right, parens=[]):
          if left:
              backtrack(p + '(', left-1, right)
          if right > left:
              backtrack(p + ')', left, right-1)
          if not right:
              parens.append(p)
          return parens
      return backtrack('', n, n)
    
# @lc code=end


s = Solution()
print(s.generateParenthesis(3))
